### PR TITLE
docs: add changelog

### LINKED ISSUE #1296: General project improvements for users and contributors
**Is your feature request related to a problem? Please describe.**
I would like to propose some general improvements to the project. The improvements kind of depend on each other a bit. That's why I decided to group them under one issue, so they can all be discussed together.

**Describe the solution you'd like**
The first improvement would be to [Keep a Changelog](https://keepachangelog.com/en/1.1.0/). I'm personally a big fan of **good** changelogs: ones that are 100% user facing, without jargon and not automated ([example](https://github.com/alacritty/alacritty/blob/ace987f343649ae98e5fb63cf825414855ccd86e/CHANGELOG.md)). To users it gives a clear overview of the important changes for them. For developers, it can be a nice way to keep track of the improvements to the project. It can function sort of like a diary in which all the significant changes are documented for future reference. It would be especially useful after a stable release, to document breaking changes.

As for the format of the changelog, I think the [Keep a Changelog](https://keepachangelog.com/en/1.1.0/) format is one of the better ones out there. It is clean, minimal and still allows anyone to quickly see all the code changes by clicking on the link in the title. It also neatly groups all related changes together under clear categories. A changelog which follows a specification is also easy to parse and use by other programs/services. One example of a service which uses these is Flathub, where the changelog is displayed under the program.

I know it can be a bit of a chore to maintain a changelog, but I would absolutely be willing to keep it updated. I'd rather do that than have it be automatically generated from the commits, as that includes a lot of noise and isn't user-facing at all. The specification also mentions it's not that important to update the changelog together with the changes, as long as it is updated before each release. It would of course be simpler to keep it updated together with each change, which leads me to the second improvement.

It could be nice to add contributor guidelines. Personally I kind of miss those, even though I've contributed before. It would be handy to get a general explanation on how to start (clone, make commit, make PR) for new contributors. It would also be handy to have commands which you can execute locally before making a PR to know your PR will pass CI. It would just be a section with the commands used in the `.github/workflows/ci.yml` file which contributors can enter locally if they want. Contributor guidelines could also include a section about updating the changelog with significant user-facing changes. That way the changelog would always be up to date and would stay in sync with the actual changes when commits are reverted. Otherwise it could be forgotten to revert the changelog when another commit it references is reverted.

**Additional context**
I'd be willing to make a first draft of all the changes. I just wanted to make a feature request first to discuss what would be wanted and what wouldn't be before I started working on it :slightly_smiling_face:

### PR DESCRIPTION
**Pull request by ThomasFrans**
closes #1296

### CODE REVIEW

**Comment by ThomasFrans** [comment]
I've had this in a private fork for some time since I wanted to see what it looked like but wanted to wait for input on #1296. The stupid GitHub bot keeps bugging me about outdated dependencies and I don't know how to turn it off in a private repo, so I'm just making this PR and removing that fork because the barrage of mails it's sending is driving me insane :smile:.

This was just for me to see what could be put in the changelog, but I think it looks pretty good and gives a handy overview for any users that are interested in what happened with the project over time. It's not finished yet. If this would be considered, it could be nice to add some contribution instructions that explain the contribution process and all the steps before creating a PR (formatting, `clippy` and adding user-facing changes to the changelog). I would also want to maintain the changelog but as I explained in #1296, I think it's less error prone to add the changes together with the commits that introduce them.

The changelog included with this PR isn't complete, but I don't think it's a requirement to go all the way back to `0.1.0`. As long as the changes are included from the point when the changelog was added, it should be fine, as it's mostly for keeping up to date with changes and documenting breaking stuff (especially after a `1.0.0` release).

**Comment by hrkfdn** [comment]
I like it a lot, thanks! I think adding a PR template makes sense to remind people to update the changelog (and READMEs). I added something simple, as I think complex/large PR templates can be rather intimidating. WDYT?

**Comment by ThomasFrans** [comment]
Glad to hear. I think the template should be as friction-free as possible like you say. It shouldn't scare new contributors (or maybe even first time contributors to any project). One tiny change I would make is to say what belongs in the changelog. It's not immediately clear that developer-facing changes are best left out.

**Comment by hrkfdn** [comment]
> I noticed that the PR template mentions 'requesting a review', but I don't think that's possible for normal contributors, so that might need to be removed.

True, to me opening a PR is somewhat synonymous to requesting a review, but I will simplify the wording.

**Comment by ThomasFrans** [comment]
I guess that's also fine. I just thought it would be confused with the 'request a review' feature of GitHub, which normal people don't have access to, but it might be clear enough that that's not what the template refers to.
